validators/bls.go

```diff
@@ -0,0 +1,227 @@
+	Address      types.Address
+	BLSPublicKey []byte
+}
+
+func (v *BLSValidator) String() string {
+	return fmt.Sprintf(
+		"%s:%s",
+		v.Address.String(),
+		hex.EncodeToString(v.BLSPublicKey),
+	)
+}
+
+func (v *BLSValidator) Type() ValidatorType {
+	return BLSValidatorType
+}
+
+func (v *BLSValidator) Addr() types.Address {
+	return v.Address
+}
+
+func (v *BLSValidator) Copy() Validator {
+	pubkey := make([]byte, len(v.BLSPublicKey))
+	copy(pubkey, v.BLSPublicKey)
+
+	return &BLSValidator{
+		Address:      v.Address,
+		BLSPublicKey: pubkey,
+	}
+}
+
+func (v *BLSValidator) Equal(vr Validator) bool {
+	vv, ok := vr.(*BLSValidator)
+	if !ok {
+		return false
+	}
+
+	return v.Address == vv.Address && bytes.Equal(v.BLSPublicKey, vv.BLSPublicKey)
+}
+
+func (v *BLSValidator) MarshalRLPWith(arena *fastrlp.Arena) *fastrlp.Value {
+	vv := arena.NewArray()
+
+	vv.Set(arena.NewBytes(v.Address[:]))
+	vv.Set(arena.NewBytes(v.BLSPublicKey))
+
+	return vv
+}
+
+func (v *BLSValidator) UnmarshalRLPFrom(p *fastrlp.Parser, val *fastrlp.Value) error {
+	elems, err := val.GetElems()
+	if err != nil {
+		return err
+	}
+
+	if len(elems) < 2 {
+		return fmt.Errorf("incorrect number of elements to decode BLSValidator, expected 2 but found %d", len(elems))
+	}
+
+	if err := elems[0].GetAddr(v.Address[:]); err != nil {
+		return fmt.Errorf("failed to decode Address: %w", err)
+	}
+
+	if v.BLSPublicKey, err = elems[1].GetBytes(v.BLSPublicKey); err != nil {
+		return fmt.Errorf("failed to decode BLSPublicKey: %w", err)
+	}
+
+	return nil
+}
+
+func (v *BLSValidator) Bytes() []byte {
+	return types.MarshalRLPTo(v.MarshalRLPWith, nil)
+}
+
+func (v *BLSValidator) SetFromBytes(input []byte) error {
+	return types.UnmarshalRlp(v.UnmarshalRLPFrom, input)
+}
+
+type BLSValidators []*BLSValidator
+
+func (vs *BLSValidators) Type() ValidatorType {
+	return BLSValidatorType
+}
+
+func (vs *BLSValidators) Len() int {
+	return len(*vs)
+}
+
+func (vs *BLSValidators) Equal(ts Validators) bool {
+	vts, ok := ts.(*BLSValidators)
+	if !ok {
+		return false
+	}
+
+	if vs.Len() != vts.Len() {
+		return false
+	}
+
+	for idx, vsv := range *vs {
+		tsv := (*vts)[idx]
+		if !vsv.Equal(tsv) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (vs *BLSValidators) Copy() Validators {
+	clone := make(BLSValidators, vs.Len())
+
+	for idx, val := range *vs {
+		clone[idx], _ = val.Copy().(*BLSValidator)
+	}
+
+	return &clone
+}
+
+func (vs *BLSValidators) At(index uint64) Validator {
+	return (*vs)[index]
+}
+
+func (vs *BLSValidators) Index(addr types.Address) int64 {
+	for i, v := range *vs {
+		if v.Address == addr {
+			return int64(i)
+		}
+	}
+
+	return -1
+}
+
+func (vs *BLSValidators) Includes(addr types.Address) bool {
+	return vs.Index(addr) != -1
+}
+
+func (vs *BLSValidators) Add(v Validator) error {
+	validator, ok := v.(*BLSValidator)
+	if !ok {
+		return ErrMismatchValidatorType
+	}
+
+	if vs.Includes(validator.Address) {
+		return ErrValidatorAlreadyExists
+	}
```

Comment:
Should this be a hard error?

Reply:
Could you tell me what hard error is and what your suggestion is?

Reply:
Does it need to error out?

Reply:
@zivkovicmilos Yes. Why do you think so?

Reply:
Resolving this comment as I have no idea where this is now, and the context under which the comment was made, since you've significantly changed the look of this PR since.